pkg/providers/tinkerbell/comparator.go

```diff
@@ -0,0 +1,51 @@
+package tinkerbell
+
+import (
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+)
+
+// VersionComparable enables kubernetes version retrieval.
+type VersionComparable interface {
+	Version() string
+}
+
+// MachineReplicaComparable enables replica number retrieval.
+type MachineReplicaComparable interface {
+	Replicas() int
+}
+
+// KubeadmControlPlaneComparator wraps the KubeadmControlPlane to enable Comparator functions.
+type KubeadmControlPlaneComparator struct {
+	*controlplanev1.KubeadmControlPlane
+}
+
+// Version retrieves KubeadmControlPlane replicas.
+func (k *KubeadmControlPlaneComparator) Version() string {
+	return k.Spec.Version
+}
+
+// Replicas retrieves KubeadmControlPlane replicas.
+func (k *KubeadmControlPlaneComparator) Replicas() int {
+	return int(*k.Spec.Replicas)
+}
+
+// MachineDeploymentComparator implements MachineReplicaComparable by decorating a MachineDeployment.
+type MachineDeploymentComparator struct {
+	*clusterv1.MachineDeployment
+}
+
+// Replicas retrieves MachineDeployment replicas.
+func (m *MachineDeploymentComparator) Replicas() int {
+	return int(*m.Spec.Replicas)
+}
+
+// HasKubernetesVersionChange detects kubernetes version has changed between two Comparator objects.
+func HasKubernetesVersionChange(oldObj VersionComparable, newObj VersionComparable) bool {
+	return oldObj.Version() != newObj.Version()
+}
+
+// ReplicasDiff returns the difference in replica count between new and old Comparator objects.
+func ReplicasDiff(oldObj MachineReplicaComparable, newObj MachineReplicaComparable) int {
```

Comment:
```suggestion
func ReplicasDiff(current MachineReplicaComparable, desired MachineReplicaComparable) int {
```

Reply:
I think we should double down on this terminology in our code!